# Model Interpretation and Insights

This notebook explores what the trained Random Forest classifier has learned from the Fake Job Postings dataset.

Rather than focusing only on prediction accuracy, this notebook analyzes the vocabulary learned by the TF-IDF vectorizer, discusses model behaviour, and identifies opportunities for future improvements.

## Objectives

- Explore the learned vocabulary
- Understand important textual patterns
- Analyze example predictions
- Discuss strengths and limitations
- Suggest future improvements

## Load Saved Model and TF-IDF Vectorizer

In [44]:
import joblib

rf_model = joblib.load("../models/fraud_detector.pkl")
tfidf = joblib.load("../models/tfidf_vectorizer.pkl")

## TF-IDF Vocabulary

The TF-IDF vectorizer converts textual information into numerical features.

Each feature represents a unique word learned during training.

In [45]:
feature_names = tfidf.get_feature_names_out()

print("Vocabulary Size:", len(feature_names))

Vocabulary Size: 10000


In [46]:
feature_names[:50]

array(['aaa', 'aan', 'ab', 'abap', 'abc', 'aberdeen', 'abilities',
       'abilitiesexperience', 'ability', 'able', 'above', 'abreast',
       'abroad', 'absence', 'absolute', 'absolutely', 'abstract', 'abuse',
       'abzorba', 'ac', 'academia', 'academic', 'academy', 'accel',
       'accelerate', 'accelerating', 'acceleration', 'accelerator',
       'accept', 'acceptable', 'acceptance', 'accepted', 'accepting',
       'access', 'accessibility', 'accessible', 'accessing',
       'accessories', 'accident', 'accidental', 'accidents', 'accion',
       'acclaimed', 'accolades', 'accommodate', 'accommodation',
       'accommodations', 'accommodationsability',
       'accommodationswillingness', 'accompanied'], dtype=object)

## Sample Learned Vocabulary

Below are a few words captured by the TF-IDF vectorizer.

In [47]:
import random

random.sample(list(feature_names), 30)

['soma',
 'weekplenty',
 'entries',
 'furnishings',
 'goes',
 'evidence',
 'enrichingarea',
 'say',
 'converting',
 'visibility',
 'driveways',
 'locals',
 'sanctuary',
 'customers',
 'relationshipswe',
 'employers',
 'vibrant',
 'motor',
 'instructions',
 'jd',
 'disciplinary',
 'possesses',
 'compensation',
 'profit',
 'revise',
 'preexisting',
 'circle',
 'del',
 'targets',
 'stated']

## Example Fraudulent Job Posting

In [ ]:
fraud_examples = df[df["fraudulent"] == 1]

fraud_examples[["title","description"]].head(3)

,title,description
98,IC&E Technician,"IC&amp;E Technician | Bakersfield, CA Mt. Poso..."
144,Forward Cap.,The group has raised a fund for the purchase o...
173,Technician Instrument & Controls,Technician Instrument &amp; ControlsLocation D...


## Example Genuine Job Posting

In [ ]:
genuine_examples = df[df["fraudulent"] == 0]

genuine_examples[["title","description"]].head(3)

,title,description
0,Marketing Intern,"Food52, a fast-growing, James Beard Award-winn..."
1,Customer Service - Cloud Video Production,Organised - Focused - Vibrant - Awesome!Do you...
2,Commissioning Machinery Assistant (CMA),"Our client, located in Houston, is actively se..."


## Observations

By comparing genuine and fraudulent job postings, several recurring patterns become apparent.

Fraudulent postings frequently contain:

- unrealistic salary claims
- urgency-based language
- requests for payment
- vague company information
- external communication channels

Genuine postings generally provide:

- detailed job descriptions
- company information
- clear responsibilities
- realistic hiring processes

## Model Strengths

- Achieved approximately 98% accuracy using Random Forest.

- Effectively distinguishes fraudulent and genuine postings.

- Fast inference time.

- Easily deployable through FastAPI.

- Supports real-time job analysis in RecruTap.

## Model Limitations

Although the model performs well, several limitations remain.

- It relies primarily on textual content.

- Images and company logos are not analyzed.

- Company reputation is not verified.

- External company websites are not validated.

- Predictions may be affected by unusual wording.

## Future Improvements

Potential future enhancements include:

- Transformer-based language models (BERT)

- Explainable AI using SHAP or LIME

- Company reputation verification

- Website trust analysis

- Salary anomaly detection

- Continuous retraining with new job postings

# Conclusion

The Random Forest classifier demonstrated strong performance in detecting fraudulent job postings using TF-IDF textual features.

The trained model has been successfully integrated into the RecruTap web application, enabling real-time fraud detection.

Future work will focus on improving explainability, expanding data sources, and enhancing the overall intelligence of the system.